### PR TITLE
feat(config): Add mergeable `addLabels` option

### LINKED ISSUE #6503: Append labels
### Which Renovate are you using?

Renovate Open Source

### Which platform are you using?

GitLab self-hosted

### What would you like to do?

I would like to append labels to merge requests. Currently labels will be overwritten completely. For example, I would like to add labels depending on language and update type. But I dont want to write a lot of rules depending on their combined status.

```
 "packageRules": [
    {
      "updateTypes": ["major"],
      "labels": ["renovate::major"]
    },
    {
      "updateTypes": ["minor"],
      "labels": ["renovate::minor"]
    },
    {
      "updateTypes": ["pin"],
      "labels": ["renovate::pin"]
    },
    {
      "languages": ["javascript"],
      "labels": ["languages::JavaScript"]
    },
    {
      "languages": ["java"],
      "labels": ["languages::Java"]
    },
    {
      "languages": ["golang"],
      "labels": ["languages::GoLang"]
    },
    {
      "languages": ["ruby"],
      "labels": ["languages::Ruby"]
    },
    {
      "languages": ["swift"],
      "labels": ["languages::Swift"]
    }
  ]
```

### PR DESCRIPTION
**Pull request by fernandomora**
## Changes:

Adds a new `addLabels` configuration option that works in the same way than existing `labels` one but is has `mergeable` tag enable.
Current labels are fully overwritten by last matched rule and combining different rules to produce a set of combined labels produces a combinatorial explosion to manually combine them.
This will allow to append labels to PRs not manually writting a lot of combined rules and labels

## Context:

Closes #6503

## Documentation (please check one with an [x])

- [X] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [X] Newly added unit tests, or
- [ ] Unit tests + ran on a real repository

### CODE REVIEW

File: docs/usage/configuration-options.md [hunk 1]
```diff
@@ -899,6 +899,30 @@ Consider this example:
 
 With the above config, every PR raised by Renovate will have the label `dependencies` while PRs containing `eslint`-related packages will instead have the label `linting`.
 
+## addLabels
```
Comment (viceice) [inline docs/usage/configuration-options.md:902]: needs to be ordered alphabetically

**Review by viceice — CHANGES_REQUESTED** [review]

File: docs/usage/configuration-options.md [hunk 2]
```diff
@@ -27,6 +27,31 @@ Also, be sure to check out Renovate's [shareable config presets](/config-presets
 
 If you have any questions about the below config options, or would like to get help/feedback about a config, please post it as an issue in [renovatebot/config-help](https://github.com/renovatebot/config-help) where we will do our best to answer your question.
 
+## addLabels
+
+Different sets of package matching `labels` are not appended to PRs, last matched rule just fully override labels to set.
```
Comment (rarkins) [inline docs/usage/configuration-options.md:32]:
```suggestion
The `labels` field is non-mergeable, meaning that any config setting a list of PR labels will replace any existing list that existed.
```

File: docs/usage/configuration-options.md [hunk 3]
```diff
@@ -27,6 +27,31 @@ Also, be sure to check out Renovate's [shareable config presets](/config-presets
...
+
+Consider this example:
+
+```json
+{
+  "labels": ["dependencies"],
+  "packageRules": [
+    {
+      "packagePatterns": ["eslint"],
+      "addLabels": ["linting"]
```
Comment (rarkins) [inline docs/usage/configuration-options.md:43]:
```suggestion
      "labels": ["linting"]
```

File: docs/usage/configuration-options.md [hunk 4]
```diff
@@ -27,6 +27,31 @@ Also, be sure to check out Renovate's [shareable config presets](/config-presets
...
+    },
+    {
+      "depTypeList": ["optionalDependencies"],
+      "addLabels": ["optional"]
+    }
+  ]
+}
+```
+
+With the above config, every PR raised by Renovate will have the label `dependencies` while PRs containing `eslint`-related packages will instead have labels `dependencies` and `linting`, PRs for optional dependencies will have `dependencies` and `optional`, and so on for different combinations of matched rules.
```
Comment (rarkins) [inline docs/usage/configuration-options.md:53]:
```suggestion
With the above config:
 - Optional dependencies will have the labels `dependencies` and `optional`
 - ESLint dependencies will have the label `linting`
 - All other dependencies will have the label `dependencies`
```

File: lib/workers/pr/index.spec.ts [hunk 5]
```diff
@@ -643,5 +643,17 @@ describe('workers/pr', () => {
         gitLabAutomerge: true,
       });
     });
+
+    it('should create a PR with set of labels and mergeable addLabels', async () => {
+      config.labels = ['deps', 'renovate'];
+      config.addLabels = ['deps', 'js'];
```
Comment (rarkins) [inline lib/workers/pr/index.spec.ts:649]:
```suggestion
```

**Review by rarkins — CHANGES_REQUESTED** [review]

File: lib/workers/pr/index.spec.ts (hunk 5 above)
Comment (fernandomora) [reply lib/workers/pr/index.spec.ts:649]: Ooops, I forgot to remove these lines. Thanks :)

Comment (fernandomora) [reply lib/workers/pr/index.spec.ts]: It seems if no global config object is mofidied this tests fails, not sure why.
I think I will probably go back and remove the new config object to use the global mutable one 🤷‍♂️

File: docs/usage/configuration-options.md [hunk 6]
```diff
@@ -27,6 +27,35 @@ Also, be sure to check out Renovate's [shareable config presets](/config-presets
 
 If you have any questions about the below config options, or would like to get help/feedback about a config, please post it as an issue in [renovatebot/config-help](https://github.com/renovatebot/config-help) where we will do our best to answer your question.
 
+## addLabels
+
+The `labels` field is non-mergeable, meaning that any config setting a list of PR labels will replace any existing list that existed.
```
Comment (HonkingGoose) [inline docs/usage/configuration-options.md:32]:
```suggestion
The `labels` field is non-mergeable, meaning that any config setting a list of PR labels will replace any existing list.
```

File: docs/usage/configuration-options.md [hunk 7]
```diff
@@ -27,6 +27,35 @@ Also, be sure to check out Renovate's [shareable config presets](/config-presets
...
+If you want to append labels for matched rules, then define an `addLabels` array of one or more label strings to append to the PR so all matched `addLabels` will be attached to PR.
```
Comment (HonkingGoose) [inline docs/usage/configuration-options.md:33]:
```suggestion
If you want to append labels for matched rules, then define an `addLabels` array with one (or more) label strings.
All matched `addLabels` strings will be attached to the PR.
```

**Review by HonkingGoose — CHANGES_REQUESTED** [review]
Got some small comments, looks good otherwise.

**Review by HonkingGoose — APPROVED** [review]

**Comment by fernandomora** [comment]
Tested in a real repository:

```
  "labels": ["dependencies"],
  "packageRules": [
    {"managers": ["sbt"], "addLabels": ["scala"]},
    {"depTypeList": ["test"],"addLabels": ["dependencies:test"]},
    {"depTypeList": ["plugin"], "addLabels": ["dependencies:plugin"]}
  ]
```

image

**Review by viceice — APPROVED** [review]

**Pull request merged**